Documentation/proposals/202405-agent-daemonset.md

```diff
@@ -0,0 +1,128 @@
+* Automatic scalability: When new nodes are added to the cluster, new Prometheus Agent pods will be automatically added in those nodes. Users can also select which set of nodes they want to deploy Prometheus Agent and the priority of Prometheus Agent pods compared to other pods on the same node.
+
+DaemonSet deployment is especially more suitable for Prometheus Agent than Prometheus, since the Agent mode requires (currently around 20 to 30%) less memory, does not produce TSDB blocks on disks, and naturally blocks querying APIs.
+
+This deployment mode has been implemented and proven in the Google Cloud Kubernetes Engine (GKE) with [Google Cloud Managed Service for Prometheus (GMP)'s operator](https://github.com/GoogleCloudPlatform/prometheus-engine/), so we can learn on their cases and collaborate on shared improvements together.
+
+
+## 2. Pitfalls of the current solution
+
+The key pitfall of managing load distribution and scalability with the current StatefulSet deployment is they are done on the cluster scope. In other words, since one or several high-availability Prometheus Agents are responsible for scraping metrics of the whole cluster, users need to calculate/estimate the load and scalability of the whole cluster to decide on replicas and sharding strategies. Estimating cluster-wide load and scalability is a much harder task than estimating node-wide load and scalability. Even though users can use helping tools like Horizontal Pod Autoscaler (HPA), that's still additional complexity.
+
+We're not saying that DaemonSet is superior to StatefulSet. In fact, StatefulSet has its own advantages, such as easier storage handling. What we're trying to say is DaemonSet can solve some of the existing problems in StatefulSet, and vice versa. So DaemonSet is a good deployment option to implement besides StatefulSet.
+
+
+## 3. Audience
+
+Users with use cases where scraped load is very large or hard to estimate and/or scalability is hard to predict, so they need an easy way to manage the load distribution and scalability.
+
+
+## 4. Goals
+
+Provide an MVP version of the DaemonSet deployment of Prometheus Agent to the Audience.
+In specific, the MVP will need to:
+* Allow users to deploy one Prometheus Agent pod per node.
+* Allow users to restrict which set of nodes they want to deploy Prometheus Agent, if desired.
+* Allow users to set the priority of Prometheus Agent pod compared to other pods on the same node, if desired.
+* Allow each Prometheus Agent pod to only scrape from the pods from PodMonitor that run on the same node as the Agent.
+
+
+## 5. Non-Goals
```

Comment:
Maybe add here that the proposal only covers the PrometheusAgent CRD because this is where we see most of the use cases. I'm not 100% sure that folks won't ask for a daemonset Prometheus CRD and maybe there are valid scenarios for it but I'd rather experiment first with the agent.